components/plugins/lcs.js

```diff
@@ -0,0 +1,758 @@
+import { inputLabelStyles } from '@brightspace-ui/core/components/inputs/input-label-styles.js';
+import { inputStyles } from '@brightspace-ui/core/components/inputs/input-styles.js';
+import { live } from 'lit-html/directives/live.js';
+import { radioStyles } from '@brightspace-ui/core/components/inputs/input-radio-styles.js';
+import { repeat } from 'lit-html/directives/repeat.js';
+import { requestInstance } from '@brightspace-ui/core/mixins/provider-mixin.js';
+import { selectStyles } from '@brightspace-ui/core/components/inputs/input-select-styles.js';
+
+const name = 'd2l-element-editor-dialog';
+
+export function addElementEditorDialogPlugin() {
+	tinymce.PluginManager.add(name, function(editor) {
+		const getSelectedElement = () => {
+			const contextNode = (editor.selection ? editor.selection.getNode() : null);
+			if (!contextNode) return null;
+			if (contextNode.tagName !== 'DIV' || !contextNode.classList.contains('d2l-element') || !contextNode.classList.contains('mceNonEditable')) return null;
+			return contextNode;
+		};
+
+		const launchElementEditor = () => {
+			const context = requestInstance(editor.getElement(), 'context');
+			const component = context.component;
+
+			let dialog = component.shadowRoot.querySelector('d2l-element-editor-dialog');
+			if (!dialog) dialog = component.shadowRoot.appendChild(document.createElement('d2l-element-editor-dialog'));
+
+			const contextNode = getSelectedElement();
+			if (contextNode) {
+				dialog.elementHTML = contextNode.outerHTML;
+			}
+
+			dialog.opened = true;
+
+			dialog.addEventListener('d2l-element-editor-dialog-close', e => {
+				if (!e.detail || !e.detail.html) return;
+				if (contextNode) {
+					editor.selection.select(contextNode);
+				}
+				editor.insertContent(e.detail.html);
+			}, { once: true });
+		};
+
+		editor.on('SetContent', function(e) {
+			const editorDocument = new DOMParser().parseFromString(e.target.getContent(), 'text/html');
+
+			if (editorDocument.querySelectorAll('.d2l-element').length > 0) {
+				let requiresContentUpdate = false;
+
+				if (styleSheets) {
+					styleSheets.forEach(styleSheet => {
+						if (!editorDocument.head.querySelector(`link[href="${styleSheet.href}"]`)) {
+							requiresContentUpdate = true;
+							const elem = editorDocument.createElement('link');
+							elem.rel = 'stylesheet';
+							elem.href = styleSheet.href;
+							elem.crossOrigin = 'anonymous';
+							editorDocument.head.appendChild(elem);
+						}
+					});
+				}
+				if (scripts) {
+					scripts.forEach(script => {
+						if (!editorDocument.body.querySelector(`script[src="${script.src}"]`)) {
+							requiresContentUpdate = true;
+							const elem = editorDocument.createElement('script');
+							elem.src = script.src;
+							elem.crossOrigin = 'anonymous';
+							editorDocument.body.appendChild(elem);
+						}
+					});
+				}
+				if (requiresContentUpdate) e.target.setContent(editorDocument.documentElement.outerHTML);
```

Comment:
Thing 3: Not so much broken as could be simplified.
```suggestion
				if (requiresContentUpdate) e.content = editorDocument.documentElement.outerHTML;
```